Measure raw event size

danielmitterdorfer opened the pull request:
With this commit we measure raw event size and add it as a new field in
the generated documents.

dliappis commented on eventdata/elasticlogs-index.json:
```diff
@@ -12,6 +12,9 @@
         "type": "date"
       },
+      "_raw_event_size": {
+        "type": "short"
```
Are we sure `raw_event` won't ever exceed 32767 characters? Looking at the [message sample](https://github.com/elastic/rally-tracks/tree/master/eventdata#example-document) seems unlikely but not sure how large the `request` string can get.

danielmitterdorfer replied:
In my tests I saw event sizes up to ~ 1000 bytes so `short` should suffice.

dliappis commented on eventdata/parameter_sources/randomevent.py:
```diff
@@ -278,7 +280,20 @@ def generate_event(self):
+                                                                         event["agent"])
+            raw_event_size = len(raw_event)
+        else:
+            raw_event_size = 0
```
I was curious to see whether going through if/else in the existing cases (where `record_raw_event_size` isn't used) affects performance; using a quick `cProfile` based microbenchmark I observed a consistent 1 sec improvement moving `raw_event_size=0` before if and removing `else` with 200million iterations. I guess we can classify this as drop in the ocean?

Conversely when `record_raw_event_size is True` and `raw_event_size = 0` is defined before `if`, therefore reset inside if, it was ~1sec slower, with the same amount of iterations compared to this code.

danielmitterdorfer replied:
Thanks for double-checking. I'd also consider this a drop in the ocean. Also, the new approach works a bit different anyway (we have completely different code paths to create the document now in order to ensure we don't have any complex code on the hot code path).

dliappis commented on eventdata/parameter_sources/randomevent.py:
```diff
@@ -260,6 +260,8 @@ def __init__(self, params):
             for d in params['delete_fields']:
                 self._delete_fields.append(d.split('.'))
 
+        self.record_raw_event_size = params.get("record_raw_event_size", True)
```
If we set this to `True` we will impact all existing benchmarks by having an additional field. Should this default to `False`?

dliappis commented on eventdata/parameter_sources/randomevent.py:
```diff
@@ -278,7 +280,20 @@ def generate_event(self):
+            raw_event_size = 0
+
         line = '{"@timestamp": "%s", ' \
+               '"_raw_event_size":%d, ' \
```
Similarly I am skeptical about having this new field (even with size `0`) for all existing benchmarks.
At the same time, I am aware if were to conditionally build the string depending on the flag, code would be messy and likely have a small perf hit.

danielmitterdorfer commented:
@dliappis I've addressed your comments and the new field is now completely optional. As a consequence I had to expose it as a track parameter in every bulk operation (thus the large changeset). Can you please have another look?

dliappis commented on eventdata/operations/indexing.json:
```diff
@@ -10,82 +10,44 @@ rolledover_indices_suffix_separator: used by the `delete_rolledover_index_patter
 {% set p_disk_type = disk_type | default('ssd') | lower %}
+
+{% set comma = joiner() %}
+{% for bulk_size in [50000, 20000, 10000, 5000, 2000, 1000, 500, 250, 125] %}
```
Nice!
